gem-check/gem-check

```diff
@@ -0,0 +1,109 @@
+#!/bin/sh
+# shellcheck disable=SC2317,SC2166,SC3043,SC2162,SC2086
+set +x
+set -f
+
+PROG="gem-check"
+
+info() {
+	echo "INFO[$PROG]:" "$@"
+}
+
+error() {
+	echo "ERROR[$PROG]:" "$@"
+	exit 1
+}
+
+fail() {
+	echo "FAIL[$PROG]:" "$@"
+	FAILS=$((FAILS+1))
+}
+
+pass() {
+	echo "PASS[$PROG]:" "$@"
+	PASSES=$((PASSES+1))
+}
+
+packages=""
+require=""
+VERBOSE=false
+
+while [ $# -ne 0 ]; do
+	case "$1" in
+		--packages=*)
+			if [ "${1#*=}" = "none" ]; then
+				packages=""
+			else
+				packages="${packages} ${1#*=}"
+			fi
+		;;
+		--packages)
+			if [ "$2" = "none" ]; then
+				packages=""
+			else
+				packages="${packages} $2"
+			fi
+			shift
+		;;
+		--require=*)
+			if [ "${1#*=}" = "none" ]; then
+				require=""
+			else
+				require="${require} ${1#*=}"
+			fi
+		;;
+		--verbose=*)
+			VERBOSE=${1#*=}
+		;;
+		--verbose)
+			VERBOSE=$2
+			shift
+		;;
+        	--*)
+			error "Unknown argument '$1'"
+		;;
+	esac
+	shift
+done
+packages=${packages# }
+require=${require# }
+
+case "$VERBOSE" in
+	true|false)
+		:
+	;;
+	*)
+		error "--verbose must be 'true' or 'false'. found '$VERBOSE'"
+	;;
+esac
+
+[ -n "${packages}" ] || error "No files or packages specified"
+
+FAILS=0
+PASSES=0
+set -- $packages
+for pkg in "$@"; do
+	gem_dash=$(echo "$pkg" | sed -e "s/^ruby[0-9\.]\+-//")
+	gem_slash=$(echo "$pkg" | sed -e "s/^ruby[0-9\.]\+-//" -e "s:\-:/:g")
+	if [ $(gem list -i "^$gem_dash$") = "true" ]; then
+		pass "Success listing gem [$gem_dash]"
+	else
+		fail "Failed listing gem [$gem_dash]"
+	fi
+	if [ -z "$require" ]; then
+		require="$gem_slash"
+	fi
+	for req in "$require"; do
+		if ruby -e "require '$req'"; then
```

Comment:
similar here, can we do

```bash
test_require() {
    local r="$1" out="" rc=""
    out=$(ruby -e '
#!/usr/bin/ruby
r = ARGV[0]
begin
  require r
  print "pass"
  exit 0
rescue LoadError => error
  print "require #{r}: #{error}"
  exit 3
end
exit 9
' "$r" 2>&1)
    rc=$?

    case "$rc:$out" in
        0:pass) return 0;;
        3:*) echo "$out"; return 1;;
    esac
    error "unexpected error in ruby require test for '$r' [$rc]: $out"
    return 1
}
```

then call it like:

```bash
if test_require "$r"; then
   pass "..."
else
   fail "..."
fi
```

Reply:
Done.